pkgs/development/python-modules/aws-sam-cli/default.nix

```diff
@@ -1,21 +1,36 @@
 { lib
-, python3
 , enableTelemetry ? false
+, buildPythonPackage
+, fetchPypi
+, aws-lambda-builders
+, aws-sam-translator
+, chevron
+, click
+, cookiecutter
+, dateparser
+, python-dateutil
+, docker
+, flask
+, jmespath
+, requests
+, serverlessrepo
+, tomlkit
+, watchdog
 }:
 
-python3.pkgs.buildPythonApplication rec {
+buildPythonPackage rec {
```

Comment:
Just realize that the consequences of having this is in `python-modules` is that it will never be able to pin any dependencies. Which can make it unusable.

my info on it:

Pinning within python-modules is highly discouraged. This will potentially introduce incompatible versions in a user's environment as either the new pinned version or the original version will be imported at runtime, breaking one of the packages.

The preference to handling this is to relax version bounds in the "install_requires" field. (could be in setup.py, pyproject.toml, requirements or others). In most cases, packages are still compatible with small API changes which may warrant a major version bump. We use test suites to verify that the package still works correctly. 

If the package is still incompatible with the latest major version, then the most proper way to handle this is make an issue with the upstream package to adopt the latest major version. Or if upstream is not very responsive, you are free patch the source to make it compatible.

In very few circumstances, two versions of the same package are allowed to exist if the packages are extremely difficult to package. Some examples of this are tensorflow, which has huge ecosystems built around it and is hard to package. Another is django, which has 2 actively developed versions, and large ecosystems built around each.

One exception to this is applications, due to the way `buildPythonApplication` and `toPythonApplication` functions work, the related derivations will not bleed dependencies between packages. If the package doesn't need to be imported by other python modules, then this package would be a good candidate to convert into application. You can look at https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/admin/awscli/default.nix as an example of using an overlay within a python application. 

Info on `buildPythonApplication` can be found [here](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#buildpythonapplication-function).
Info on `toPythonApplication` can be found [here](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#topythonapplication-function).

Reply:
aws (at least in the past) has been very "pin happy", and generally is a monumental PITA to get it to play nicely with other modules.

Reply:
Thanks for all the background info.

Certainly awscli looks difficult to package, and [awscli2 even more so](https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/admin/awscli2/default.nix#L3-L22), but maybe aws-sam-cli is a little better (?). It currently doesn't have anything pinned, only a few loosened requirements, and we already have many of its first-party dependencies (i.e. ones written by the AWS devs) in python-modules, like `aws-sam-translator`, `serverlessrepo`, and `aws-lambda-builders`, so I figured throwing in `aws-sam-cli` itself wouldn't be a very drastic change.

If things go south with later versions of `aws-sam-cli`, we could always move it back out of python-modules, right?

Reply:
yea, but it takes time and effort to do so. And there's 60,000+ other packages to maintain.

AFAIK, `aws-sam-cli` is intended to be used as a cli application, and may not intend for users to use it as a python module.

Reply:
Unfortunately, `nix-shell` appears to run the shellHooks, which includes setting the `$PYTHONPATH`... which is causing you some issues.